Species specific footprints for footwear/clothing

Nerezza opened the pull request:
Unathi footwear is usually depicted as fitting around their digitigrade feet and the large claws on them, however even with their claws usually poking out of their footwear they've still left human-shaped footprints behind.

This PR implements a feature for species-specific footprints for footwear. Hardsuit boots on a human or skrell leaves human-shaped footprints, while hardsuit boots on an Unathi leaves Unathi-shaped footprints. This also works for magboots, and Unathi footwear.

Only one item exists that has unique footprint decals. It has been updated to work with the new proc.

The new proc is hooked in for use by all clothing instead of just footwear because footprints are also taken from space suits too, since they also cover the feet.

Testing image below, the same pair of magboots were passed between a Human and an Unathi repeatedly to make sure the correct footprints were chosen every time. Not pictured is the HCM test (passed) and making sure the footwear with unique footprints still worked correctly (passed).
![image](https://user-images.githubusercontent.com/1986009/212843410-fc9a86cc-3d1b-4e98-afc4-376f2e497b7d.png)
No maintenance drones were harmed in the making of this PR.

:cl: Nerezza
rscadd: Unathi now leave their unique clawprints even while wearing footwear.
/:cl:
<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You'll find a README and example file in .\html\changelogs\ for further instructions.

You can also find a template for adding your changelog directly to the PR description here: https://github.com/Baystation12/Baystation12/wiki/Automatic-changelog-generation
-->

Nerezza committed:
Add species specific footwear tracks

Change footwear tracks decals according to the species wearing the footwear.

Nerezza committed:
Footprint changes comments

Forgot to finish comments before my last commit

SierraKomodo commented on code/modules/clothing/_clothing.dm:
```diff
@@ -16,7 +16,14 @@
 	var/volume_multiplier = 1
 
-	var/move_trail = /obj/effect/decal/cleanable/blood/tracks/footprints // if this item covers the feet, the footprints it should leave
+	// The footprints left by items covering the feet (space suits and shoes) chosen by wearer bodytype (NOT SPECIES).
+	// "Default" is used if the wearer's bodytype isn't in the list.
+	// Change species_trails to give custom footprints, move_trail doesn't need to be changed..
+	var/list/species_trails = list(
+		"Default" = /obj/effect/decal/cleanable/blood/tracks/footprints,
+		SPECIES_UNATHI = /obj/effect/decal/cleanable/blood/tracks/claw
+	)
```
Comments for vars such as this should use the dmdoc formatting. This allows the VSC extension and the documentation generator to parse and include this documentation in popups, the website, etc.

https://github.com/Baystation12/Baystation12/wiki/Guide-to-SpacemanDMM-Docblocks

Nerezza replied:
Roger, didn't know this was a thing now. I'll fix it after I have some coffee.

SierraKomodo commented on code/modules/clothing/_clothing.dm:
```diff
@@ -98,10 +105,23 @@
 		update_vision()
 	return ..()
 
+/obj/item/clothing/proc/species_tracks(mob/user)
+	//Change the tracks decal according to species.
```
As above regarding dmdoc formatting

https://github.com/Baystation12/Baystation12/wiki/Guide-to-SpacemanDMM-Docblocks

Nerezza committed:
SpacemanDMM comments

Rewrites comments for SpacemanDMM parsing. Two other unrelated comment edits for increased clarity.
Also removes a defunct line I forgot to remove from sneaky laceups.

Nerezza commented:
I've redone the comments to work with SpacemanDMM o7

Spookerton commented on code/modules/clothing/_clothing.dm:
```diff
@@ -98,10 +114,23 @@
 	return 1
 
 /obj/item/clothing/equipped(mob/user)
+	species_tracks(user)
```
```suggestion
	update_species_tracks(user)
```

Spookerton commented on code/modules/clothing/_clothing.dm:
```diff
@@ -98,10 +114,23 @@
 		update_vision()
 	return ..()
 
+/// This proc is called when a clothing item is equipped to determine what footprints it gives the wearer. See species_trails for details on how to set this up.
+/obj/item/clothing/proc/species_tracks(mob/user)
+	var/bodytype = "Default"					// Should never try to find a null bodytype.
+	var/mob/living/carbon/human/H
+	if(ishuman(user))
+		H = user
+		bodytype = H.species.get_bodytype()		// Gets the species bodytype, this is NOT the user's species.
+	if(!species_trails[bodytype])				// If the species bodytype isn't assigned a specific footprint, use the assigned default.
+		move_trail = species_trails["Default"]
+	else										// Otherwise use the footprint assigned to the bodytype.
+		move_trail = species_trails[bodytype]
+
```
```suggestion
/// Set move_trail according to the current wearer's species.
/obj/item/clothing/proc/update_species_tracks(mob/living/carbon/human/user)
	if (!species_trails)
		return
	if (istype(user))
		move_trail = species_trails[user.species.get_bodytype()]
		if (move_trail)
			return
	move_trail = species_trails[SPECIES_HUMAN]
```

Spookerton commented on code/modules/clothing/_clothing.dm:
```diff
@@ -16,7 +16,23 @@
+		"Default" = /obj/effect/decal/cleanable/blood/tracks/footprints,
+		SPECIES_UNATHI = /obj/effect/decal/cleanable/blood/tracks/claw
+	)
+	/// Use species_trails instead. This variable is now only used to store the footprint decals chosen when someone equips a clothing item.
+	var/move_trail = /obj/effect/decal/cleanable/blood/tracks/footprints // Default value set since there's no reason for this to be null.
```
```suggestion
	/// The effect to create when moving on filth.
	var/move_trail = /obj/effect/decal/cleanable/blood/tracks/footprints
```

Spookerton commented on code/modules/clothing/_clothing.dm:
```diff
@@ -16,7 +16,23 @@
 	var/volume_multiplier = 1
 
-	var/move_trail = /obj/effect/decal/cleanable/blood/tracks/footprints // if this item covers the feet, the footprints it should leave
+	/**
+	* The footprints decal left by items covering the feet. This includes space suits, not just footwear.
+	* This is called by species_tracks() to select the correct decal by wearer body type which is not the same as their species.
+	* Calling get_bodytype() on SPECIES_YEOSA returns SPECIES_UNATHI as an example.
+	*
+	* '"Default"' - Always include this. This can be the only value in the list if every species should use the specified footprint decals.
+	* If the wearer's body type doesn't match any of the specified body types, this is the value the proc uses by default.
+	*
+	* 'SPECIES_UNATHI' - An example of a valid body type. Assign any body types a custom footprint decal and the proc will use the specified decal instead of the default.
+	* Multiple body types can have specified footprint decals in the list.
+	*/
+	var/list/species_trails = list(
+		"Default" = /obj/effect/decal/cleanable/blood/tracks/footprints,
+		SPECIES_UNATHI = /obj/effect/decal/cleanable/blood/tracks/claw
+	)
```
```suggestion
	/// If set, a (bodytype = /obj) path map for updating move_trail. Always supply SPECIES_HUMAN.
	var/list/species_trails
```

Spookerton replied:
nb here you would also want to set on `/obj/item/clothing/shoes`
```
species_trails = list(
		SPECIES_HUMAN = /obj/effect/decal/cleanable/blood/tracks/footprints,
		SPECIES_UNATHI = /obj/effect/decal/cleanable/blood/tracks/claw
	)
```

Spookerton commented on code/modules/clothing/shoes/miscellaneous.dm:
```diff
@@ -230,7 +230,9 @@
 	desc = "The height of fashion, and they're pre-polished. Upon further inspection, the soles appear to be on backwards. They look uncomfortable."
 	species_restricted = list(SPECIES_HUMAN, SPECIES_IPC)
-	move_trail = /obj/effect/decal/cleanable/blood/tracks/footprints/reversed
+	species_trails = list(
+		"Default" = /obj/effect/decal/cleanable/blood/tracks/footprints/reversed
+	)
```
```suggestion
	move_trail = /obj/effect/decal/cleanable/blood/tracks/footprints/reversed
```

Spookerton commented:
Note that this behavior even after CRs are resolved will be ungood - it creates a unique list on every shoes item for an effectively static value. A better solution would be to leverage the existing `/datum/species/proc/get_move_trail` with a per-species path list for overrides checked in order of highest specificity first, like:

```
/datum/species
  var/list/shoe_trail_overrides

/datum/species/unathi
  shoe_trail_overrides = list(
    /obj/item/clothing/shoes/laceup/sneakies = /obj/effect/decal/cleanable/blood/tracks/footprints/reversed,
    /obj/item/clothing/shoes = /obj/effect/decal/cleanable/blood/tracks/claw
  )
```

and then in `/datum/species/proc/get_move_trail` replacing `return shoes.move_trail`
```
for (var/key in shoe_trail_overrides)
  if (istype(shoes, key))
    return shoe_trail_overrides[key]
return shoes.move_trail
```

Nerezza commented:
Yesterday morning I remembered initial(var) exists which would simplify this even more.

Today I had more time to think about things and realized things like:
1. Clothing types like glasses don't need a list of footwear sprites to pick from.
2. The vast majority of shoes are only wearable by species that have or mimic plantigrade feet and would not need alternate species footprints assigned.
3. The only suit slot items needing alternate footprints are ones that cover the feet and can be worn by non-plantigrade species, the definitions for which are already a mess and adding a new variable to them would just be inherently messy.
4. I wrote this at like 3am in the morning so in hindsight no surprise I didn't realize some glaringly obvious issues with this solution.

tl;dr: Spookerton didn't review this at 3am so his solution here is great. I'll close this PR to keep the dev branch commit history clean and open another with a /datum/species based alternative once I'm done testing it.
